File: src/Legi.Catalog.Domain/ValueObjects/Isbn.cs
using Legi.SharedKernel;

namespace Legi.Catalog.Domain.ValueObjects;

public sealed class Isbn : ValueObject
{
    public string Value { get; }

    private Isbn(string value)
    {
        Value = value;
    }

    public static Isbn Create(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new DomainException("ISBN is required");

        // Remove hyphens and spaces
        var cleaned = value.Replace("-", "").Replace(" ", "").Trim().ToUpperInvariant();

        switch (cleaned.Length)
        {
            case 10:
            {
                if (!IsValidIsbn10(cleaned))
                    throw new DomainException("Invalid ISBN-10 checksum");
                break;
            }
            case 13:
            {
                if (!IsValidIsbn13(cleaned))
                    throw new DomainException("Invalid ISBN-13 checksum");
                break;
            }
            default:
                throw new DomainException("ISBN must be 10 or 13 characters");
        }

        return new Isbn(cleaned);
    }

    /// <summary>
    /// Validates ISBN-10 checksum.
    /// Formula: (10*d1 + 9*d2 + 8*d3 + ... + 1*d10) mod 11 == 0
    /// Last digit can be 'X' representing 10.
    /// </summary>
    private static bool IsValidIsbn10(string isbn)
    {
        if (isbn.Length != 10)
            return false;

        var sum = 0;
        for (var i = 0; i < 10; i++)
        {
            int digit;
            if (i == 9 && isbn[i] == 'X')
            {
                digit = 10;
            }
            else if (char.IsDigit(isbn[i]))
            {
                digit = isbn[i] - '0';
            }
            else
            {
                return false;
            }

            sum += (10 - i) * digit;
        }

        return sum % 11 == 0;
    }

    /// <summary>
    /// Validates ISBN-13 checksum.
    /// Formula: (d1 + 3*d2 + d3 + 3*d4 + ... + d13) mod 10 == 0
    /// </summary>
    private static bool IsValidIsbn13(string isbn)
    {
        if (isbn.Length != 13)
            return false;

        if (!isbn.All(char.IsDigit))
            return false;

        var sum = 0;
        for (var i = 0; i < 13; i++)
        {
            var digit = isbn[i] - '0';
            sum += (i % 2 == 0) ? digit : digit * 3;
        }

        return sum % 10 == 0;
    }

    protected override IEnumerable<object> GetEqualityComponents()
    {
        yield return Value;
    }

    public override string ToString() => Value;

    public static implicit operator string(Isbn isbn) => isbn.Value;
}

File: web/legi-web/src/features/catalog/lib/isbn.ts
export function normalizeIsbn(value: string) {
  return value.replace(/[-\s]/g, "").trim().toUpperCase();
}

export function isValidIsbn(value: string) {
  const isbn = normalizeIsbn(value);

  if (isbn.length === 10) {
    return isValidIsbn10(isbn);
  }

  if (isbn.length === 13) {
    return isValidIsbn13(isbn);
  }

  return false;
}

function isValidIsbn10(isbn: string) {
  let sum = 0;

  for (let index = 0; index < isbn.length; index += 1) {
    const char = isbn[index];
    const digit = index === 9 && char === "X" ? 10 : Number(char);

    if (!Number.isInteger(digit) || digit < 0 || digit > 10 || (digit === 10 && index !== 9)) {
      return false;
    }

    sum += (10 - index) * digit;
  }

  return sum % 11 === 0;
}

function isValidIsbn13(isbn: string) {
  if (!/^\d{13}$/.test(isbn)) {
    return false;
  }

  const sum = isbn
    .split("")
    .reduce((total, char, index) => total + Number(char) * (index % 2 === 0 ? 1 : 3), 0);

  return sum % 10 === 0;
}
